sql/api/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2699,7 +2699,8 @@ abstract class Dataset[T] extends Serializable {
    * @group typedrel
    * @since 1.6.0
    */
-  def transform[U](t: Dataset[T] => Dataset[U]): Dataset[U] = t(this.asInstanceOf[Dataset[T]])
+  def transform[U, DSO[_] <: Dataset[_]](t: this.type => DSO[U]): DSO[U] =
```

Comment:
Sorry about this. I spend a lot of time fighting the type system. This makes a lot more sense.